### PR TITLE
pipewire: update to 0.3.6

### PR DESCRIPTION
**Pull request by sirn**
This PR updates `pipewire` to 0.3.6 and also updating/patching all packages that depends on it. 

* `xdg-desktop-portal` includes a patch backported from 1.7.0 to allow 1.6.0 to build against pipewire-0.3 (1.7.x is prerelease)
* `xdg-desktop-portal-kde` requires pkg-config to detect pipewire-0.3 because of include dir version suffix.
* `mutter` includes a patch backported from 3.35 to allow 3.34 to build against pipewire-0.3 (upgrading to >=3.35 also requires upgrading GNOME).

### CODE REVIEW

**Comment by Duncaen** [comment]
Those xdg-desktop-portal updates are pre releases and conditional patches are big no go.

**Comment by sirn** [comment]
Oh, I didn't realized it was a pre-release. I'll try to upstream the patch (given that glibc has `O_SYNC` aliased to `O_FSYNC`). Should I close this for the time being? (e.g. until xdg-desktop-portal has a pipewire-0.3 compatible release)

**Comment by Piraty** [comment]
@sirn you can prefix the PR's title with `WIP: ` (preferred), or `NOMERGE` or mark it as draft

**Comment by sirn** [comment]
I've bumped `xdg-desktop-portal` to 1.7.2 which include a fix for building under musl so conditional patch is no longer needed. It's still prerelease though, but I'll keep this PR updated until 1.8.0(?) is released. :bowing_man:

**Comment by mnabid** [comment]
@sirn
I rebuilt `xdg-desktop-portal` 1.6.0 with [this patch](https://github.com/flatpak/xdg-desktop-portal/commit/a38901e5e7f835efe7b7a06c55790c8c20bc91a2) and that seems to be built fine. 

God knows when **1.8.0** will be released ! `mutter-3.36` needs `libpipewire-0.3.x `, so all **gnome 3.36** updates will be stuck if pipewire can't be updated. Could you just revbump `xdg-desktop-portal` for now? I've modified [template](https://github.com/mnabid/void-packages/commit/a91e2a2a8177ad231f1f49f168b44279f5f18ec0) a bit.

**Comment by sirn** [comment]
@mnabid I've cherry-pick pipewire from your branch. Thank you!

**Comment by sirn** [comment]
@Duncaen since this PR is now building `xdg-desktop-portal` 1.6.0 with pipewire 0.3 patch, and it seems to build in all systems (& also confirmed working locally on x86_64-musl), is it possible to re-review? Thank you! :bow:

**Comment by mnabid** [comment]
Now that I think about it...
`mutter-3.34` is still in the repo and needs `libpipewire-0.2`, so that'll create **unresolvable shlib** issue. I'm working on gnome 3.36 updates (#21382).

**Comment by sirn** [comment]
Looks like so far only mutter hasn't been updated:

```
$ xbps-query -RX libpipewire
gstreamer1-pipewire-0.2.7_1
mutter-3.34.1_2
pipewire-0.2.7_1
pipewire-devel-0.2.7_1
xdg-desktop-portal-1.6.0_1
xdg-desktop-portal-kde-5.18.4.1_1
```

Is it possible to version bump just Mutter, or that requires the rest of GNOME 3.36 as well? In the latter case I guess it makes more sense that I'll close this PR and let #21382 incorporate pipewire-0.3?

**Comment by mnabid** [comment]
> Looks like so far only mutter hasn't been updated:
> 
> ```
> $ xbps-query -RX libpipewire
> gstreamer1-pipewire-0.2.7_1
> mutter-3.34.1_2
> pipewire-0.2.7_1
> pipewire-devel-0.2.7_1
> xdg-desktop-portal-1.6.0_1
> xdg-desktop-portal-kde-5.18.4.1_1
> ```
> 
> Is it possible to version bump just Mutter, or that requires the rest of GNOME 3.36 as well? In the latter case I guess it makes more sense that I'll close this PR and let #21382 incorporate pipewire-0.3?

Yeah, rest of 3.36.

**Comment by sirn** [comment]
A patch at https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/1062 does seems to get mutter to built against pipewire 0.3. I'm not sure if I need to revbump something else. @mnabid do you mind testing if screencast under this patched mutter-3.34 works?

**Comment by mnabid** [comment]
Yeah, works fine. Thanks !

File: srcpkgs/pipewire/template [hunk 1]
```diff
@@ -1,52 +1,65 @@
...
 homepage="https://pipewire.org/"
-changelog="https://raw.githubusercontent.com/PipeWire/pipewire/master/NEWS"
-distfiles="https://github.com/PipeWire/pipewire/archive/${version}.tar.gz"
-checksum=bfaa0f6ae6c0791e2e0b59234d399753bf24f1b33dbf587682363a8463dd8df1
+changelog="https://gitlab.freedesktop.org/pipewire/pipewire/-/raw/master/NEWS"
+distfiles="https://gitlab.freedesktop.org/pipewire/pipewire/-/archive/${version}/pipewire-${version}.tar.gz"
+checksum=fd1bb2ae65997d921d6c18e1828f23251dac3145bbe5031c324afeb39f9b26fb
 conf_files="/etc/pipewire/pipewire.conf"
 
+CFLAGS+="-Wformat"
```
Comment (Piraty) [inline srcpkgs/pipewire/template:21]: append without a leading whitespace doesn't make sense (template is sourced multiple times). why add this at all?

Comment (sirn) [reply srcpkgs/pipewire/template:21]: This was due to pipewire has `-Wformat-security` by default, but GCC will complain/ignore the flag if `-Wformat` wasn't also set, but I guess it's doesn't really matter. I'm removing it.

**Comment by sirn** [comment]
Updated to pipewire-0.3.4 which needs removing `-Werror=suggest-attribute=format` from `meson.build` (enabled since 0.3.3). See also: https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/233

pipewire-0.3.4 also now provides `pw-pulse` and `pw-jack` which is a wrapper for setting `LD_LIBRARY_PATH` to pipewire's pulseaudio/JACK implementation for anyone to try out without replacing their system library. This is part of `pipewire-{pulseaudio,jack}` subtree so I install them in `libpulseaudio-pipewire` and `libjack-pipewire` respectively.

**Comment by sirn** [comment]
Replaced `-Wformat` and `-Werror=suggest-attribute=format` with a patch from upstream which now fixes warnings/compilation on musl. armv6l/armv7l currently failing due to [neon support](https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/235).

File: srcpkgs/xdg-desktop-portal-kde/template [hunk 2]
```diff
@@ -1,7 +1,7 @@
 # Template file for 'xdg-desktop-portal-kde'
 pkgname=xdg-desktop-portal-kde
 version=5.18.4.1
-revision=1
+revision=2
```
Comment (mnabid) [inline srcpkgs/xdg-desktop-portal-kde/template:4]: **5.18.5** is in the repo.

Comment (sirn) [reply srcpkgs/xdg-desktop-portal-kde/template]: Updated

**Comment by sirn** [comment]
Compilation on armv6l/armv7l is now fixed with a [patch from upstream](https://gitlab.freedesktop.org/pipewire/pipewire/-/commit/011992e774fcb58966f567099f03943c57e66c2b). I'm removing WIP (for now...).

**Comment by mnabid** [comment]
> Compilation on armv6l/armv7l is now fixed with a [patch from upstream](https://gitlab.freedesktop.org/pipewire/pipewire/-/commit/011992e774fcb58966f567099f03943c57e66c2b). I'm removing WIP (for now...).

I tried that patch. No luck.

**Comment by not-chicken** [comment]
@sirn  It's still failing on arm* : (

**Comment by sirn** [comment]
@not-chicken I noticed this also, but still haven't got a chance to take a detailed look. Strange that the NEON check program compiles but actual build says unsupported :\ Will take a look over the weekend.

**Comment by mnabid** [comment]
@sirn As far as I can remember, `pipewire-0.3.2` was the last one still working for `arm*`. So I tried reverting [`a81e269a`](https://gitlab.freedesktop.org/pipewire/pipewire/-/commit/a81e269a7e058f97f333a1a5c2f393ecbbb5b3dd) (partially, only **meson.build**) and I was able to build for arm again. 

Here's the [patch](https://github.com/mnabid/void-packages/commit/45537fd5e6db1e7b08018a6b21b2c46df5a27975#diff-6761ede86a57277d05036e9d5cc36295).

File: common/shlibs [hunk 3]
```diff
@@ -3204,7 +3204,7 @@ libarcan_shmif_ext.so.0.11 arcan-0.5.4.3_1
 libarcan_shmif_server.so.0.11 arcan-0.5.4.3_1
 libarcan_tui.so.0.11 arcan-0.5.4.3_1
 liblwipv6.so.2 lwipv6-1.5a_1
-libpipewire-0.2.so.1 libpipewire-0.2.2_1
+libpipewire-0.3.so.0 libpipewire-0.3.4_1
```
Comment (ifreund) [inline common/shlibs:3207]:
```suggestion
libpipewire-0.3.so.0 libpipewire-0.3.5_1
```

Comment (ifreund) [reply common/shlibs:3207]: also, shouldn't the other shared libraries (`libspa` etc) show up here?

Comment (ifreund) [reply common/shlibs:3207]: poked at this a bit more, all of the following `.so` files are currently included in the `pipewire` package and likely need to be added to `common/shlibs` and move to appropriate packages
```
/usr/lib/spa-0.2/alsa/libspa-alsa.so
/usr/lib/spa-0.2/audioconvert/libspa-audioconvert.so
/usr/lib/spa-0.2/audiomixer/libspa-audiomixer.so
/usr/lib/spa-0.2/control/libspa-control.so
/usr/lib/spa-0.2/jack/libspa-jack.so
/usr/lib/spa-0.2/support/libspa-dbus.so
/usr/lib/spa-0.2/support/libspa-support.so
/usr/lib/spa-0.2/v4l2/libspa-v4l2.so
/usr/lib/spa-0.2/videoconvert/libspa-videoconvert.so
/usr/lib/spa-0.2/vulkan/libspa-vulkan.so
```

Comment (sirn) [reply common/shlibs:3207]: I don't think libspa are supposed to be linked (e.g. added to shlibs), at least for `alsa`, `jack`, `v4l2` and `vulkan`, as they're API plugins meant to be consumed by pipewire itself.

I considered moving alsa/jack/v4l2 and vulkan to a subpackage (just like `libspa-ffmpeg` and `libspa-bluez`) but v4l2 and alsa are actually included in `pipewire-0.2.7`. Not sure what would be the best way to do this, maybe with some defaults pipewire depends?

**Comment by sirn** [comment]
> @sirn As far as I can remember, `pipewire-0.3.2` was the last one still working for `arm*`. So I tried reverting [`a81e269a`](https://gitlab.freedesktop.org/pipewire/pipewire/-/commit/a81e269a7e058f97f333a1a5c2f393ecbbb5b3dd) (partially, only **meson.build**) and I was able to build for arm* again.
> 
> Here's the [patch](https://github.com/mnabid/void-packages/commit/45537fd5e6db1e7b08018a6b21b2c46df5a27975#diff-6761ede86a57277d05036e9d5cc36295).

This one works! Though, I'm starting to wonder if we should try to enable NEON for non-`aarch64` at all, given xbps is distributing a binary package and not all armv6l/armv7l has NEON as it would probably going to segfault in some armv6l/armv7l systems if pipewire were built with NEON support by default.

**Comment by shizonic** [comment]
Since https://github.com/void-linux/void-packages/pull/21382 depends on this may I ask if there is still one working on? AFAIK it's the small little missing piece for the GNOME 3.36.x update. So I would really appreciate if someone focus on again.

Of course If I can help, please let me know.

**Comment by sirn** [comment]
I'm still working on it.

**Comment by shizonic** [comment]
> I'm still working on it.

Well thank you!

**Comment by sirn** [comment]
Updated to include a patch from @mnabid & split libspa into its own package per @ifreund.

**Comment by not-chicken** [comment]
I don't know anything about arm, neon, etc but unless the detection fails due to something void specific maybe it's worth reopening https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/235 
In the meantime, I think it's fine to roll with the patch.

**Comment by sirn** [comment]
Now failing due to ICU... probably because of recent ICU update missing some revbump?

**Comment by not-chicken** [comment]
> Now failing due to ICU... probably because of recent ICU update missing some revbump?

Builders are still busy rebuilding stuff for ICU 67.1 https://build.voidlinux.org/waterfall
We gotta wait till they are finished to get a clean CI run.

**Comment by mnabid** [comment]
@sirn Travis should run fine now.

**Comment by sirn** [comment]
All greens!

**Comment by mnabid** [comment]
Seems like everything is awesome ! 
Pinging @Piraty for a review.

File: srcpkgs/pipewire/template [hunk 4]
```diff
@@ -1,52 +1,119 @@
...
 hostmakedepends="doxygen graphviz pkg-config xmltoman"
 makedepends="SDL2-devel ffmpeg-devel gst-plugins-base1-devel jack-devel
- sbc-devel v4l-utils-devel libva-devel"
+ sbc-devel v4l-utils-devel libva-devel libbluetooth-devel"
 short_desc="Server and user space API to deal with multimedia pipelines"
-maintainer="Orphaned <orphan@voidlinux.org>"
-#Next release will probably change to MIT, currently listed in master
-license="LGPL-2.1-only, GPL-2.0-only "
+maintainer="Kridsada Thanabulpong <sirn@ogsite.net>"
+license="MIT"
```
Comment (Piraty) [inline srcpkgs/pipewire/template:14]: seems correct. upstream's commit mentions `MIT/X11` yet the additional `advertisement` clause is in fact missing, which means it's pure MIT license...

File: srcpkgs/pipewire/patches/fix-neon-detection.patch [hunk 5]
```diff
@@ -0,0 +1,52 @@
+diff --git meson.build meson.build
```
Comment (Piraty) [inline srcpkgs/pipewire/patches/fix-neon-detection.patch:1]: if this is really the revert of an upstream commit, be so kind and note that here, ideally linking upstream issue discussion (if any) for later reference

**Comment by ifreund** [comment]
Is it really preferable to patch e.g. `xdg-desktop-portal` rather than package a pre-release? Also a little confused by the versioning of `xdg-desktop-portal`, it is true that `1.6.0` is marked as a release while all the others are marked as pre-releases on their github page but I see no mention of this fact in their changelog or anywhere else in the repo. Based off the version numbers and content of the repository itself it would seem to me that `1.7.0` is a normal release.

**Comment by bobertlo** [comment]
There is a conflict now on mutter. is anyone working on this?

```
mutter-3.34.1_3: broken, unresolvable shlib `libpipewire-0.2.so.1'
```

many moons ago i was looking into upgrading mutter but this was the conflict. i could look into this again if it is not being pursued.

**Comment by sirn** [comment]
Weird, there's already a patch to fix this at https://github.com/void-linux/void-packages/commit/3b2f6c25bb2dc282a401a1ab6de8e9d2372d0dc6

**Comment by Piraty** [comment]
@bobertlo are you sure? i hit sth similar as well due to my version of mutter being manually built from testing a gnome update PR
```
○ xi mutter
[*] Updating repository `https://alpha.de.repo.voidlinux.org/current/x86_64-repodata' ...
[*] Updating repository `https://alpha.de.repo.voidlinux.org/current/multilib/x86_64-repodata' ...

Name          Action    Version           New version            Download size
zenity        install   -                 3.32.0_1               4152KB 
gnome-desktop install   -                 3.34.1_1               435KB 
libcanberra   install   -                 0.30_11                - 
libpipewire   install   -                 0.3.6_1                374KB 
mutter        install   -                 3.34.1_3               2285KB
```

**Comment by mnabid** [comment]
I installed `mutter`, `libpipewire` and `graphene` from the repos and faced no shlib conflict. 

```
[nafis@Npetune ~]$ for pkg in mutter graphene libpipewire; do xbps-query $pkg -p pkgver,repository; done
mutter-3.34.1_3
https://alpha.de.repo.voidlinux.org/current
graphene-1.8.2_2
https://alpha.de.repo.voidlinux.org/current
libpipewire-0.3.6_1
https://alpha.de.repo.voidlinux.org/current

```

**Comment by bobertlo** [comment]
Oh wow I forgot I patched mutter for a bug. I had a backport PR then decided to wait for 3.36. sorry lol

**Pull request merged**